joern-cli/frontends/javasrc2cpg/src/main/scala/io/joern/javasrc2cpg/passes/AstCreator.scala

```diff
@@ -1292,22 +1360,477 @@ class AstCreator(filename: String, javaParserAst: CompilationUnit, global: Globa
+        .code(iterableName)
+        .typeFullName(iterableType)
+        .order(order)
+        .lineNumber(lineNo)
+    val iterableLocalAst = Ast(iterableLocalNode)
+
+    val iterableAssignNode =
+      NewCall()
+        .name(Operators.assignment)
+        .methodFullName(Operators.assignment)
+        .typeFullName(iterableType)
+        .dispatchType(DispatchTypes.STATIC_DISPATCH)
+        .order(order + 1)
+        .lineNumber(lineNo)
+    val iterableAssignIdentifier =
+      NewIdentifier()
+        .name(iterableName)
+        .code(iterableName)
+        .typeFullName(iterableType)
+        .order(1)
+        .argumentIndex(1)
+        .lineNumber(lineNo)
+    val iterableAssignArgs = List(Ast(iterableAssignIdentifier), iterableAst)
+    val iterableAssignAst =
+      callAst(iterableAssignNode, iterableAssignArgs)
+        .withRefEdge(iterableAssignIdentifier, iterableLocalNode)
+
+    (iterableLocalNode, List(iterableLocalAst, iterableAssignAst))
+  }
+
+  private def nativeForEachIdxLocalNode(lineNo: Option[Integer]): NewLocal = {
+    val idxName = nextIndexName()
+    val idxLocal =
+      NewLocal()
+        .name(idxName)
+        .typeFullName(TypeConstants.Int)
+        .code(idxName)
+        .lineNumber(lineNo)
+        .order(1)
+    scopeStack.addToScope(idxName, idxLocal)
+    idxLocal
+  }
+
+  private def nativeForEachIdxInitializerAst(lineNo: Option[Integer], idxLocal: NewLocal): Ast = {
+    val idxName = idxLocal.name
+    val idxInitializerCallNode =
+      NewCall()
+        .name(Operators.assignment)
+        .methodFullName(Operators.assignment)
+        .code(s"int $idxName = 0")
+        .typeFullName(TypeConstants.Int)
+        .order(2)
+        .argumentIndex(2)
+        .lineNumber(lineNo)
+        .dispatchType(DispatchTypes.STATIC_DISPATCH)
+    val idxIdentifierArg = identifierFromNamedVarType(idxLocal, lineNo, 1)
+    val zeroLiteral =
+      NewLiteral()
+        .code("0")
+        .typeFullName(TypeConstants.Int)
+        .order(2)
+        .argumentIndex(2)
+        .lineNumber(lineNo)
+    val idxInitializerArgAsts = List(Ast(idxIdentifierArg), Ast(zeroLiteral))
+    callAst(idxInitializerCallNode, idxInitializerArgAsts)
+      .withRefEdge(idxIdentifierArg, idxLocal)
+  }
+
+  private def nativeForEachCompareAst(
+    lineNo: Option[Integer],
+    iterableLocal: NamedVariableNodeType,
+    idxLocal: NewLocal
+  ): Ast = {
+    val idxName      = idxLocal.name
+    val iterableName = iterableLocal.name
+
+    val compareNode =
+      NewCall()
+        .name(Operators.lessThan)
+        .methodFullName(Operators.lessThan)
+        .code(s"$idxName < $iterableName.length")
+        .typeFullName(TypeConstants.Boolean)
+        .dispatchType(DispatchTypes.STATIC_DISPATCH)
+        .order(3)
+        .argumentIndex(3)
+        .lineNumber(lineNo)
+    val comparisonIdxIdentifier = identifierFromNamedVarType(idxLocal, lineNo, 1)
+    val comparisonFieldAccess =
+      NewCall()
```

Comment:
if you make a fn like this one https://github.com/joernio/joern/blob/master/joern-cli/frontends/kotlin2cpg/src/main/scala/io/joern/kotlin2cpg/ast/Nodes.scala#L210-L226 you'll make the code a lot more readable. highly recommended